llvm/lib/Target/AIE/AIEEliminateDuplicatePHI.cpp

```diff
@@ -0,0 +1,160 @@
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineModuleInfo.h"
+#include "llvm/CodeGen/TargetPassConfig.h"
+#include "llvm/InitializePasses.h"
+
+#define DEBUG_TYPE "aie-duplicate-PHI-elimination"
+
+using namespace llvm;
+
+static const char AIE_ELIMINATE_DUPLICATE_PHI[] =
+    "AIE Eliminate Duplicate PHI Pass";
+
+namespace {
+
+/// Eliminates duplicate PHI uses by replacing all uses of DuplicateReg with
+/// BaseReg.
+///
+/// This function iterates over all instructions that use DuplicateReg and
+/// replaces any register operands that match DuplicateReg with BaseReg.
+///
+/// \param MRI The MachineRegisterInfo object for the current function.
+/// \param Observer GISelObserverWrapper object to observe and track changes.
+/// \param BaseReg The register to replace DuplicateReg with.
+/// \param DuplicateReg The register to be replaced.
+static void eliminateDuplicatePHIUses(MachineRegisterInfo &MRI,
+                                      GISelObserverWrapper &Observer,
+                                      Register BaseReg, Register DuplicateReg) {
+  const auto UseInstIter =
+      make_early_inc_range(MRI.use_instructions(DuplicateReg));
+  for (auto &UseInstr : UseInstIter) {
+    for (auto &MO : UseInstr.uses()) {
+      if (MO.isReg() && MO.getReg() == DuplicateReg) {
+        Observer.changingInstr(UseInstr);
+        MO.setReg(BaseReg);
+        Observer.changedInstr(UseInstr);
+      }
+    }
+  }
+}
+
+/// Returns true if the instruction is a PHI node with a pointer type operand.
+/// \param MRI The MachineRegisterInfo object for the current function.
+/// \param MI The instruction to check.
+static bool isPointerOperandPHI(MachineRegisterInfo &MRI,
+                                const MachineInstr &MI) {
+  assert(MI.getOpcode() == TargetOpcode::G_PHI);
+  const int DefaultAddrSpace = 0;
+  const Register Reg = MI.getOperand(0).getReg();
+  const LLT RegType = MRI.getType(Reg);
+  return RegType == LLT::pointer(DefaultAddrSpace, RegType.getSizeInBits());
+}
+
+class AIEEliminateDuplicatePHI : public MachineFunctionPass {
+public:
+  static char ID;
+  AIEEliminateDuplicatePHI() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MF) override {
+    MachineRegisterInfo &MRI = MF.getRegInfo();
+    TargetPassConfig &TPC = getAnalysis<TargetPassConfig>();
+    // Enable CSE to set up GISelObserverWrapper object
+    GISelCSEAnalysisWrapper &Wrapper =
+        getAnalysis<GISelCSEAnalysisWrapperPass>().getCSEWrapper();
+    auto *CSEInfo = &Wrapper.get(TPC.getCSEConfig());
+    std::unique_ptr<MachineIRBuilder> Builder =
+        CSEInfo ? std::make_unique<CSEMIRBuilder>()
+                : std::make_unique<MachineIRBuilder>();
+    Builder->setMF(MF);
+    MachineIRBuilder &MIB = *Builder;
+    // Set Observer to inform change of register type or change in op-code
+    GISelObserverWrapper Observer;
+    if (CSEInfo) {
```

Comment:
Aren't we implementing a dedicated CSE pass here? So could we merge this with another run of a CSE pass after address space lowering?

Reply:
This basically reuse the very first CSE pass created here.  For now that is in address space flattening.